### ISSUE TITLE
🟡 MEDIUM: Fix N+1 query problems in card/forum listings

### ISSUE THREAD

**Issue by techstoa** [SmolTech/CityForge#67]
## Severity
**Medium** - Performance issue

## Description
Card and forum listings have N+1 query problems, executing hundreds of queries instead of one.

## Affected Files
- `src/lib/db/queries.ts:99-122` - `getCardsByTag` loads tags in loop
- `src/lib/db/queries.ts:187-205` - `getForumThreads` loads posts separately
- `src/app/api/cards/route.ts:119` - Loads reviews for each card

## Current Code
```typescript
export async function getCards() {
  // 1 query: Get all cards
  const cards = await prisma.card.findMany();
  
  // N queries: Get tags for each card
  for (const card of cards) {
    card.tags = await prisma.tag.findMany({
      where: { cards: { some: { card_id: card.id } } }
    });
  }
  
  return cards;
}
// Result: 1 + N queries (N = number of cards)
```

## Performance Impact
- 100 cards = 101 queries
- Slow response times (100ms → 5000ms)
- High database load
- Poor scalability

### 1. Use Prisma Include
```typescript
export async function getCards() {
  // Single query with JOIN
  const cards = await prisma.card.findMany({
    include: {
      tags: true,
      reviews: {
        select: {
          id: true,
          rating: true,
          user: {
            select: {
              id: true,
              username: true
            }
          }
        }
      },
      _count: {
        select: {
          reviews: true
        }
      }
    }
  });
  
  return cards;
}
// Result: 1 query
```

### 2. Use DataLoader Pattern
```typescript
import DataLoader from 'dataloader';

const tagLoader = new DataLoader(async (cardIds: number[]) => {
  const cardTags = await prisma.cardTag.findMany({
    where: { card_id: { in: cardIds } },
    include: { tag: true }
  });
  
  // Group by card_id
  const byCardId = new Map<number, Tag[]>();
  for (const ct of cardTags) {
    if (\!byCardId.has(ct.card_id)) {
      byCardId.set(ct.card_id, []);
    }
    byCardId.get(ct.card_id)\!.push(ct.tag);
  }
  
  return cardIds.map(id => byCardId.get(id) || []);
});

// Usage
const cards = await prisma.card.findMany();
const cardsWithTags = await Promise.all(
  cards.map(async card => ({
    ...card,
    tags: await tagLoader.load(card.id)
  }))
);
```

### 3. Optimize Forum Queries
```typescript
export async function getThreadsWithLatestPost(categoryId: number) {
  const threads = await prisma.forumThread.findMany({
    where: { category_id: categoryId },
    include: {
      posts: {
        orderBy: { created_at: 'desc' },
        take: 1,
        include: {
          user: {
            select: { id: true, username: true }
          }
        }
      },
      _count: {
        select: { posts: true }
      },
      user: {
        select: { id: true, username: true }
      }
    },
    orderBy: [
      { pinned: 'desc' },
      { updated_at: 'desc' }
    ]
  });
  
  return threads;
}
```

### 1. Enable Prisma Query Logging
```typescript
const prisma = new PrismaClient({
  log: ['query', 'info', 'warn', 'error']
});
```

### 2. Use Prisma Studio
```bash
npx prisma studio
```

### 3. Monitor Database Logs
```sql
-- PostgreSQL
ALTER DATABASE cityforge SET log_statement = 'all';
```

## Priority
Month 1 - Medium priority performance issue

## Labels
enhancement, performance, database

**Comment by techstoa** [comment]
Closing as outdated - this issue references Prisma/TypeScript query patterns, but CityForge's backend is Django (Python).

For Django N+1 query optimization, use:
- select_related() for foreign keys
- prefetch_related() for reverse relations and M2M
- annotate() with Count/Avg for aggregations

If you identify specific N+1 problems in the Django views/querysets, please open a new Django-specific issue with file paths and query details.
